pkg/deploy/controller/image_change_controller.go

```diff
@@ -64,65 +66,102 @@ func (c *ImageChangeController) HandleImageRepo() {
-		glog.V(2).Infof("Error generating new version of deploymentConfig %v", configName)
-		return err
-	}
+// triggerMatchesImages decides whether a given trigger for config matches the provided image repo.
+// When matching:
+// - The trigger From field is preferred over the deprecated RepositoryName field.
+// - The namespace of the trigger is preferred over the config's namespace.
+func triggerMatchesImage(config *deployapi.DeploymentConfig, trigger *deployapi.DeploymentTriggerImageChangeParams, repo *imageapi.ImageRepository) bool {
+	if len(trigger.From.Name) > 0 {
+		namespace := trigger.From.Namespace
+		if len(namespace) == 0 {
+			namespace = config.Namespace
+		}
 
-	// update the deployment config with the trigger that resulted in the new config being generated
-	newConfig.Details = generateTriggerDetails(triggers)
+		return repo.Namespace == namespace && repo.Name == trigger.From.Name
+	}
 
-	_, err = c.DeploymentConfigInterface.UpdateDeploymentConfig(newConfig.Namespace, newConfig)
-	if err != nil {
-		glog.V(2).Infof("Error updating deploymentConfig %v", configName)
-		return err
+	// This is an invalid state (as one of From.Name or RepositoryName is required), but
+	// account for it anyway.
+	if len(trigger.RepositoryName) == 0 {
+		return false
 	}
 
-	return nil
+	// If the repo's repository information isn't yet available, we can't assume it'll match.
+	return len(repo.Status.DockerImageRepository) > 0 &&
+		trigger.RepositoryName == repo.Status.DockerImageRepository
 }
 
-func parseImage(name string) (string, string) {
-	index := strings.LastIndex(name, ":")
-	if index == -1 {
-		return "", ""
+func (c *ImageChangeController) regenerate(imageRepo *imageapi.ImageRepository, config *deployapi.DeploymentConfig, triggers []deployapi.DeploymentTriggerImageChangeParams) error {
+	// Get a regenerated config which includes the new image repo references
+	newConfig, err := c.DeploymentConfigInterface.GenerateDeploymentConfig(config.Namespace, config.Name)
+	if err != nil {
+		glog.V(2).Infof("Error generating new version of deploymentConfig %v", config.Name)
+		return err
 	}
 
-	return name[:index], name[index+1:]
-}
-
-func generateTriggerDetails(triggers []deployapi.DeploymentTriggerImageChangeParams) *deployapi.DeploymentDetails {
-	// Generate the DeploymentCause objects from each DeploymentTriggerImageChangeParams object
-	// Using separate structs to ensure flexibility in the future if these structs need to diverge
+	// Update the deployment config with the trigger that resulted in the new config
 	causes := []*deployapi.DeploymentCause{}
 	for _, trigger := range triggers {
+		repoName := trigger.RepositoryName
+
+		if len(repoName) == 0 {
+			if len(imageRepo.Status.DockerImageRepository) == 0 {
+				// If the trigger relies on a image repo reference, and we don't know what docker repo
+				// it points at, we can't build a cause for the reference yet.
+				continue
+			}
+
+			id, ok := imageRepo.Tags[trigger.Tag]
+			if !ok {
```

Comment:
Maybe. If the generator didn't return an error due to the missing tag, it could mean the copy in memory here is stale. Regardless, what's the correct response? Abort the whole operation (even though the generator successfully gave us a new config)? Generate an "unknown" or error-looking cause? Log and ignore?